### ISSUE TITLE
frequent occurrences of freezing or hanging when loading the sampler

### ISSUE THREAD

**Issue by Changyuan0825** [MITIBMxGraph/SALIENT_plusplus_artifact#4]
When running the "bash ./run_sim_experiments.sh --dataset_name ogbn-products --dataset_dir ../dataset --partition_labels_dir ../dataset/partition-labels --num_partitions 2", there are frequent occurrences of freezing or hanging when loading the sampler at line 71 of SALIENT_plusplus_artifact/caching/vip.py (i.e.,   for mfg in sampler:). May I ask how you handle such issues? Thank you!

**Comment by timkaler** [comment]
Could you share the version of pytorch geometric and pytorch you are using? Also a few details about the machine you are running on would be helpful --- e.g., the number of physical CPUs, amount of RAM, and GPU.

We didn't encounter this issue when performing the original experiments or preparing/testing the artifact. I have, however, had a master's student who recently encountered a similar issue this fall.

The "freezing" is happening when using the pytorch geometric neighborsampler. A longer term fix for this artifact is probably to replace the neighborsampler with SALIENT's fast sampler --- mainly so that the artifact can be compatible with newer versions of pytorch geometric that have made changes to the neighborsampler/neighborloader.

As for quick, potential fixes, I might suggest trying one or both of the following:

1) try modifying line 66 `num_workers=num_workers_sampler,` and change num_workers_sampler to be equal to max(2,P-1) where P is the number of physical CPUs on your machine.

2) try setting `pin_memory = False` on line 67. I do not think there is any need to use pinned memory for the simulation code.

I may not have time to look into this more carefully for 1-2 weeks, unfortunately, but we definitely want to fix this properly. As such, I'd certainly appreciate any additional information you collect about the issue and your attempts to resolve it. 

Cheers,
Tim

**Comment by Changyuan0825** [comment]
Thanks for your help! I am grateful for your response. My version of pytorch geometric and pytorch is same with you. My machine has 2 CPUs (each has 24 cores), and I think the neighborSampler is excuted at CPU, so it may not be related to the GPU. My codes performed will when I add torch.set_num_threads(1). Through I do not know why this makes snese. I referred https://discuss.pytorch.org/t/possible-deadlock-training-example-gets-stuck/94221.

Thank you very much!
Changyuan
